### PR TITLE
Make reads and point counts consistent during resharding 

### PR DESCRIPTION
**Pull request by timvisee**
Tracked in: #4213 
Depends on: #4609

This PR ensures that reads and point counting remain consistent while resharding is going on. It adds some tests to assert this behavior.

Most importantly, two changes in our read/filtering logic were required:
1. don't always apply resharding point filter, apply on _old_ shards, but not the _new_ shard
2. `select_shards` should include the _new_ shard on read hash ring commit (right after migrating points)

Tests cover stability assertions for point counts, scrolling and the query API. For each test we reshard 3 times in sequence and make sure the responses remain exactly the same.

### Tasks
- [x] Merge https://github.com/qdrant/qdrant/pull/4609
- [x] Rebase on `dev`
- [x] Undraft

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by generall** [comment]
Might be overlapping with  https://github.com/qdrant/qdrant/pull/4619

File: lib/collection/src/collection/point_ops.rs [hunk 1]
```diff
@@ -252,15 +247,36 @@ impl Collection {
...
+            let resharding_filter = shards_holder.resharding_filter();
+            let reshard_shard_id = shards_holder
+                .resharding_state
+                .read()
+                .as_ref()
+                .map(|state| state.shard_id);
+
+            // Take the filter and apply the resharding filter
+            // Should be used on all shards, except the new resharding shard
+            let mut filter = request.filter.clone();
```
[the remark this answers is not in the document]
Comment (generall) [reply lib/collection/src/collection/point_ops.rs:261]: Maybe a function which would consume the request and either return the same one or clone and modify?

[the remark this answers is not in the document]
Comment (timvisee) [reply lib/collection/src/collection/point_ops.rs]: I've now removed unnecessary cloning in two cases; it now only clones while resharding is going on.

I'm not super happy about it though. The fitment in code is a bit ugly.

**Review by ffuugoo — COMMENTED** [review]
Overall, LGTM, though, I'm not sure about point filtering in `retrieve`.

File: lib/collection/src/collection/point_ops.rs [hunk 2]
```diff
@@ -433,11 +450,6 @@ impl Collection {
         let points = all_shard_collection_results
             .into_iter()
             .flatten()
-            // If resharding is in progress, and *read* hash-ring is committed, filter-out "resharded" points
-            .filter(|point| match &resharding_filter {
-                Some(filter) => filter.check(point.id),
-                None => true,
-            })
```
[the remark this answers is not in the document]
Comment (timvisee) [reply lib/collection/src/collection/point_ops.rs:440]: I believe you're right. I think I removed the 'old filtering' and didn't put it back here.

I'll add it back and follow the pattern of the other operations.

[the remark this answers is not in the document]
Comment (timvisee) [reply lib/collection/src/collection/point_ops.rs:440]: Implemented in 1271350713f6707b928f3ba96924447a4f398190

**Review by ffuugoo — APPROVED** [review]
I'd simplify filter setup in `retrieve`, but otherwise LGTM.

File: lib/collection/src/collection/point_ops.rs [hunk 3]
```diff
@@ -419,6 +451,14 @@ impl Collection {
                         if shard_key.is_none() {
                             return Ok(records);
                         }
+
+                        // If resharding, exclude migrated points from old shards
+                        if let Some(filter) = resharding_filter.as_ref() {
+                            if do_filter {
+                                records.retain(|record| !filter.check(record.id));
+                            }
+                        }
```
Comment (ffuugoo) [inline lib/collection/src/collection/point_ops.rs:460]: Also, I'm not 100% sure, but I think we might have to filter points *before* the `shard_key.is_none()` early return.

**Comment by ffuugoo** [comment]
Here's my over-engineered solution. It spans too many lines to put it as suggestion on the review.

```rust
            // Resharding filter to apply when resharding is active
            let resharding_filter = shard_holder.resharding_filter_impl();
            let reshard_shard_id = shard_holder
                .resharding_state
                .read()
                .as_ref()
                .map(|state| state.shard_id);

            let retrieve_futures = target_shards.into_iter().map(|(shard, shard_key)| {
                // Explicitly borrow `request` and `with_payload`, so we can use them in `async move`
                // block below without unnecessarily cloning anything
                let request = &request;
                let with_payload = &with_payload;

                // If resharding, prepare filter to exclude migrated points from *old* shards
                let resharding_filter = resharding_filter
                    .as_ref()
                    .filter(|_| Some(shard.shard_id) != reshard_shard_id);

                async move {
                    let mut records = shard
                        .retrieve(request.clone(), &with_payload, &request.with_vector, read_consistency, shard_selection.is_shard_id())
                        .await?;

                    // If resharding, exclude migrated points from *old* shards
                    if let Some(filter) = resharding_filter {
                        records.retain(|record| !filter.check(record.id));
                    }

                    if shard_key.is_none() {
                        return Ok(records);
                    }

                    for point in &mut records {
                        point.shard_key.clone_from(&shard_key.cloned());
                    }

                    CollectionResult::Ok(records)
                }
            });
```

**Comment by timvisee** [comment]
> Here's my over-engineered solution. It spans too many lines to put it as suggestion on the review.

Looks good. Would you mind to add a commit to this PR?

**Comment by ffuugoo** [comment]
> Looks good. Would you mind to add a commit to this PR?

Done.

**Comment by timvisee** [comment]
> Might be overlapping with #4619

A resharding filter for the query API is apparently still missing. Implementing that will also resolve having the filter for #4619. I'll work on this tomorrow.

**Pull request merged**